src/apputils/clientsession.ts

```diff
@@ -0,0 +1,819 @@
+   * kernel name.  If no kernel has been started, this is a no-op.
+   */
+  restart(): Promise<Kernel.IKernel | null> {
+    let kernel = this.kernel;
+    if (!kernel) {
+      if (this._prevKernelName) {
+        return this.changeKernel({ name: this._prevKernelName });
+      }
+      return Promise.resolve(null);
+    }
+    return ClientSession.restartKernel(kernel);
+  }
+
+  /**
+   * Select a kernel for the session.
+   */
+  selectKernel(): Promise<void> {
+    return Private.selectKernel(this, this._manager);
+  }
+
+  /**
+   * Change the session path.
+   *
+   * @param path - The new session path.
+   *
+   * @returns A promise that resolves when the session has renamed.
+   *
+   * #### Notes
+   * This uses the Jupyter REST API, and the response is validated.
+   * The promise is fulfilled on a valid response and rejected otherwise.
+   */
+  setPath(path: string): Promise<void> {
+    this._path = path;
+    if (this._session) {
+      return this._session.rename(path);
+    }
+    return Promise.resolve(void 0);
+  }
+
+  /**
+   * Change the session name.
+   */
+  setName(name: string): Promise<void> {
+    // no-op until supported.
+    this._name = name;
+    return Promise.resolve(void 0);
+  }
+
+  /**
+   * Change the session type.
+   */
+  setType(type: string): Promise<void> {
+    // no-op until supported.
+    this._type = type;
+    return Promise.resolve(void 0);
+  }
+
+  /**
+   * Start a session and set up its signals.
+   */
+  private _startSession(model: Kernel.IModel): Promise<Kernel.IKernel> {
+    this._promise = new PromiseDelegate<void>();
+    return this._manager.startNew({
+      path: this._path,
+      kernelName: model.name,
+      kernelId: model.id
+    }).then(session => {
+      return this._handleNewSession(session);
+    }).catch(err => {
+      return this._handleSessionError(err);
+    });
+  }
+
+  /**
+   * Handle a new session object.
+   */
+  private _handleNewSession(session: Session.ISession): Kernel.IKernel {
+    if (this.isDisposed) {
+      return null;
+    }
+    if (this._session) {
+      this._session.dispose();
+    }
+    this._promise.resolve(void 0);
+    this._promise = null;
+    this._session = session;
+    session.terminated.connect(this._onTerminated, this);
+    session.pathChanged.connect(this._onPathChanged, this);
+    session.kernelChanged.connect(this._onKernelChanged, this);
+    session.statusChanged.connect(this._onStatusChanged, this);
+    session.iopubMessage.connect(this._onIopubMessage, this);
+    session.unhandledMessage.connect(this._onUnhandledMessage, this);
+    this._kernelChanged.emit(session.kernel);
+    this._prevKernelName = session.kernel.name;
+    return session.kernel;
+  }
+
+  /**
+   * Handle an error in session startup.
+   */
+  private _handleSessionError(err: utils.IAjaxError): Promise<void> {
+    this._promise.resolve(void 0);
+    this._promise = null;
+    let response = JSON.parse(err.xhr.response);
```

Comment:
This can `throw` (if the server is returning something that is not well-formed). In general, all calls to `JSON.parse()` that rely on foreign data should probably be in a `try ... catch` block.